fix: missing index and geodata import process uses normal table

zackpollard opened the pull request:
This PR:
- Aligns the index names with what is in the database migrations
- Re-adds the missing alternateNames index, this was slowing location search down to ~1.5 seconds from 50ms on my prod instance.
- Removes any need for tmp tables outside of just copying the initial table structure
- Removes any possibility of lingering tmp tables
- Adds the indexes to the new table definitions in code

zackpollard committed:
chore: add geodata indexes to table definitions

zackpollard committed:
chore: rename incorrectly name geodata index

mertalev commented on server/src/repositories/map.repository.ts:
```diff
@@ -311,28 +309,33 @@ export class MapRepository {
   }
 
-  private createGeodataIndices() {
+  private createGeodataIndices(manager: Transaction<MapDB>) {
```
Using the same connection for all of them might be a lot slower. Can you test if there's a difference?

jrasm91 reviewed (approved):
LGTM. Did you use the migration script to generate the migration?

mertalev commented:
I just compared and the index creation + renaming took 6219.80ms vs 3320.56ms on main. I don't think we should change that part.

zackpollard commented:
> I just compared and the index creation + renaming took 6219.80ms vs 3320.56ms on main. I don't think we should change that part.

It's slower but it means we can name everything correctly. We don't want drift from what is in the table definitions. Not doing it in a transaction would mean that failure could cause problems. Although, we should have the server just outright fail to start on a failure anyway, so we could probably ditch the transaction entirely.

mertalev commented:
You could probably move the `createGeodataIndices` call to after the renaming transaction and let it still use multiple connections.

zackpollard commented:
> You could probably move the `createGeodataIndices` call to after the renaming transaction and let it still use multiple connections.

At that point why even use transactions, you can still leave the database in a broken state. I'll change it tomorrow so that any failure causes the server to fail to startup, I don't know why it isn't like this already honestly...

mertalev commented:
> > You could probably move the `createGeodataIndices` call to after the renaming transaction and let it still use multiple connections.
> 
> At that point why even use transactions, you can still leave the database in a broken state. I'll change it tomorrow so that any failure causes the server to fail to startup, I don't know why it isn't like this already honestly... Right now something can fail and just leave the server in a broken state.

It isn't like that right now because the swap happens at the very end, so the ingestion only takes effect if the entire thing succeeds. If it fails, the next boot drops any existing temp table and it's like the previous failure never happened. The renaming is more sensitive because you don't want it to be possible for `geodata_places` to go missing, etc.

zackpollard commented:
> > > You could probably move the `createGeodataIndices` call to after the renaming transaction and let it still use multiple connections.
> > 
> > 
> > At that point why even use transactions, you can still leave the database in a broken state. I'll change it tomorrow so that any failure causes the server to fail to startup, I don't know why it isn't like this already honestly... Right now something can fail and just leave the server in a broken state.
> 
> It isn't like that right now because the swap happens at the very end, so the ingestion only takes effect if the entire thing succeeds. If it fails, the next boot drops any existing temp table and it's like the previous failure never happened. The renaming is more sensitive because you don't want it to be possible for `geodata_places` to go missing, etc.

Yep, so it can just fail silently forever and nobody ever knows about it 😆

mertalev commented:
Wouldn't the server crash on an error? I was assuming that it wouldn't start up if it failed since this is part of the init bootstrapping.

zackpollard commented:
> Wouldn't the server crash on an error? I was assuming that it wouldn't start up if it failed since this is part of the init bootstrapping.

In metadata.service it is wrapped in a try catch with no re-throw

mertalev commented:
Ohhhhhh, gotcha. Agree it shouldn't start up if it fails here. I thought that was already how it works lol

zackpollard committed:
fix: import into geodata places with correct index names

zackpollard commented:
@mertalev I have made some changes. The indexes are created outside of the transaction again now, the only thing done within the transaction is to create the tmp table, drop the old table and rename the tmp table afterwards. I removed the initial check to drop the tmp table if it exists because that should now be impossible given we run this in a transaction. To cleanup any hanging tmp tables, I added a "drop if exists" to the migrations.

mertalev reviewed (approved):
Nice!
